dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/security/spring2fa/TwoFactorAuthenticationProvider.java

```diff
@@ -103,22 +108,21 @@ public Authentication authenticate( Authentication auth )
             throw new BadCredentialsException( "Invalid login method, user is using external authentication." );
         }
 
-        validateTwoFactor( user, details );
+        validateTwoFactor( user, auth.getDetails() );
 
         return new UsernamePasswordAuthenticationToken( userService.createUserDetails( user ),
             result.getCredentials(),
             result.getAuthorities() );
     }
 
-    private void validateTwoFactor( User user, ForwardedIpAwareWebAuthenticationDetails details )
+    private void validateTwoFactor( User user, Object details )
```

Comment:
Is there no common base class or interface you could use instead of `Object` ?

Reply:
Unfortunately not, sometimes as with OAuth it is a map, other cases it is WebAuthenticationDetails